File: arrays/268_Missing_Number.cpp
/**
 * @file    268_Missing_Number.cpp
 *
 * @author  Juan Diego Becerra
 * @date    2021-11-20
 * @brief   https://leetcode.com/problems/missing-number/
 *
 */

#include <vector>
using namespace std;

// Time: O(n) | Space: O(1)
int missingNumber(vector<int> &nums)
{
    int n, expected_sum, actual_sum;
    actual_sum = 0;
    n = nums.size();
    expected_sum = (n * (n + 1)) / 2; // sum from [1, n]

    for (int num : nums)
    {
        actual_sum += num;
    }
    return expected_sum - actual_sum;
}

File: arrays/268_Missing_Number.py
"""268_Missing_Number.py

Author: Juan Diego Becerra
Date:   06-08-2022
Brief:  https://leetcode.com/problems/missing-number/
"""


# Time: O(n) | Space: O(1)
def missingNumber(nums: list[int]) -> int:
    n = len(nums)
    expected_sum = (n * (n + 1)) // 2  # sum of consecutive numbers from 0 to n
    real_sum = 0

    for num in nums:
        real_sum += num

    return expected_sum - real_sum
